### DISCUSSION TITLE
Portable Expression 2 Concept

### DISCUSSION THREAD

**Discussion by dstmnaki** [wiremod/wire#2976] [Suggestions]
Portable Expression 2 is a swep counter part to the real life flipper zero

It's basicly a viewmodel that has multiple buttons, an egp screen for graphics and a cable for connecting to sockets (if that works)

The buttons would only function if you are holding the swep, same thing with the cable, should you switch to another swep it would detach the cable

E2 functions:
-pe2OpenMenu() - opens the editor
-pe2SendAnimation(Animation_name string) - plays a specified viewmodel animation using string
-pe2SendAnimation(Animation_id number) - plays a specified viewmodel animation using number
-pe2PlugToSocket() - if owner is looking at a socket and is in a reasonable distance, create a wire cable that connects from the swep entity to the socket entity
-pe2UnPlug() - if the cable is plugged, this can be called to unplug the cable

for the egp part one could just use entity():wirelink():egp...in case you don't have plans on adding entirely new functions for it, it would also make this much faster to code

Console commands:
-pe2OpenMenu - opens the editor in case you remove the function from the e2

Example code

```
@name Portable Expression 2

event keyPressed(Player:entity, Key:string, Down:number, KeyBind:string) {
    if(Player==owner()&Down){
        switch(Key){
            case "r",
                pe2OpenMenu()
            break
            
            case "mouse_left",
                pe2SendAnimation("button_1")
            break
            
            case "mouse_right",
                if(IsPlugged){
                    pe2UnPlug()
                }else{
                    pe2PlugToSocket()
                }
            break
        }
    }
}
```

**Comment by dstmnaki** [comment]
i should probably also mention that the pe2SendAnimation functions would be used to play animations that use the buttons etc, to make it look more alive

**Comment by vurvdev** [comment]
This seems too specific to add, and the names are pretty confusing, do you just want toolgun functionality with E2? I made an extension in my VExtensions addon for that if you want it. It could probably be ported to wire with little issues

**Comment by dstmnaki** [comment]
ok i get everything else but how are the names confusing? pe2 is just
portable expression 2 but shortened...

ke 10. tammik. 2024 klo 0.01 Vurv ***@***.***) kirjoitti:

**Comment by dstmnaki** [comment]
after checking vextensions, i think you misinterpreted the entire idea of
what i was going for

ke 10. tammik. 2024 klo 6.13 Lauri Jokimäki ***@***.***)
kirjoitti:

**Comment by vurvdev** [comment]
Anyway my point still stands about this being way too specific. This is an addons job to do. You could do it entirely as a separate addon.
If something like this were to be included in base wiremod it should be more general.

I believe you could get this behavior with the wire laser pointer, or just using a hands swep and manually handling keypresses and eye traces. You could also get the behavior with my vextensions toolgun which is why I brought it up.

Or if people really like the idea, at the very least, there shouldn't be "god functions" like `pe2PlugToSocket` which relies entirely on the user having the swep out already and aiming at a socket (so there's pretty much no point in the E2 existing at that point, and it's just worse for the user than them clicking to do it themself)
